OrbitCaptureClient/GpuQueueSubmissionProcessor.cpp

```diff
@@ -0,0 +1,398 @@
+
+  if (matching_gpu_submission->meta_info().pre_submission_cpu_timestamp() > submit_time) {
+    return nullptr;
+  }
+
+  return matching_gpu_submission;
+}
+
+const GpuJob* GpuQueueSubmissionProcessor::FindMatchingGpuJob(
+    int32_t thread_id, uint64_t pre_submission_cpu_timestamp,
+    uint64_t post_submission_cpu_timestamp) {
+  const auto& submission_time_to_gpu_job_it = tid_to_submission_time_to_gpu_job_.find(thread_id);
+  if (submission_time_to_gpu_job_it == tid_to_submission_time_to_gpu_job_.end()) {
+    return nullptr;
+  }
+
+  const auto& submission_time_to_gpu_job = submission_time_to_gpu_job_it->second;
+
+  // Find the first Gpu job that has a timestamp greater or equal to the "pre submission" timestamp:
+  auto gpu_job_matching_pre_submission_it =
+      submission_time_to_gpu_job.lower_bound(pre_submission_cpu_timestamp);
+  if (gpu_job_matching_pre_submission_it == submission_time_to_gpu_job.end()) {
+    return nullptr;
+  }
+
+  // Find the first Gpu job that has a timestamp greater to the "post submission" timestamp
+  // (which would be the next job) and decrease the iterator by one.
+  auto gpu_job_matching_post_submission_it =
+      submission_time_to_gpu_job.upper_bound(post_submission_cpu_timestamp);
+  if (gpu_job_matching_post_submission_it == submission_time_to_gpu_job.begin()) {
+    return nullptr;
+  }
+  --gpu_job_matching_post_submission_it;
+
+  if (&gpu_job_matching_pre_submission_it->second != &gpu_job_matching_post_submission_it->second) {
+    return nullptr;
+  }
+
+  return &gpu_job_matching_pre_submission_it->second;
+}
+
+std::vector<TimerInfo> GpuQueueSubmissionProcessor::ProcessGpuQueueSubmissionWithMatchingGpuJob(
+    const GpuQueueSubmission& gpu_queue_submission, const GpuJob& matching_gpu_job,
+    const absl::flat_hash_map<uint64_t, std::string>& string_intern_pool,
+    const std::function<uint64_t(const std::string& str)>&
+        get_string_hash_and_send_to_listener_if_necessary) {
+  std::vector<TimerInfo> result;
+  std::string timeline;
+  if (matching_gpu_job.timeline_or_key_case() == GpuJob::kTimelineKey) {
+    CHECK(string_intern_pool.contains(matching_gpu_job.timeline_key()));
+    timeline = string_intern_pool.at(matching_gpu_job.timeline_key());
+  } else {
+    timeline = matching_gpu_job.timeline();
+  }
+  uint64_t timeline_hash = get_string_hash_and_send_to_listener_if_necessary(timeline);
+
+  std::optional<GpuCommandBuffer> first_command_buffer =
+      ExtractFirstCommandBuffer(gpu_queue_submission);
+
+  ProcessGpuCommandBuffers(gpu_queue_submission, matching_gpu_job, first_command_buffer,
+                           timeline_hash, get_string_hash_and_send_to_listener_if_necessary,
+                           &result);
+
+  ProcessGpuDebugMarkers(gpu_queue_submission, matching_gpu_job, first_command_buffer, timeline,
+                         string_intern_pool, get_string_hash_and_send_to_listener_if_necessary,
+                         &result);
```

Comment:
Wouldn't it be a clearer if these methods returned a `std::vector<TimerInfo>` as well and these were then appended to `result`? Or is it because of performance?

Reply:
There was no special reason. Changed it.